File: sdap-python/src/sdap/crypto/canonicalize.py
"""JSON Canonical Serialization (JCS) — RFC 8785."""

from __future__ import annotations

import json
import math


def canonicalize(obj: dict) -> bytes:
    """Return RFC 8785 canonical JSON bytes for *obj*.

    Rules:
    - Keys sorted lexicographically (Unicode code-point order), recursively
    - No insignificant whitespace
    - Strings encoded as UTF-8 with standard JSON escaping
    - Numbers: no trailing zeros in fractions; integers have no decimal point;
      special floats (inf, nan) are not permitted
    - null → ``null``, booleans → ``true`` / ``false``
    """
    return _serialize(obj).encode("utf-8")


def _serialize(value: object) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        if math.isnan(value) or math.isinf(value):
            raise ValueError("JCS does not support NaN or Infinity")
        # Use Python's repr-level precision, strip unnecessary trailing zeros
        return _serialize_float(value)
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False)
    if isinstance(value, (list, tuple)):
        items = ",".join(_serialize(v) for v in value)
        return f"[{items}]"
    if isinstance(value, dict):
        pairs = ",".join(
            f"{json.dumps(k, ensure_ascii=False)}:{_serialize(v)}"
            for k, v in sorted(value.items())
        )
        return "{" + pairs + "}"
    raise TypeError(f"Unsupported type for JCS: {type(value)}")


def _serialize_float(value: float) -> str:
    """Serialize a float following ES2019 / RFC 8785 numeric rules."""
    # Use repr for round-trip fidelity; strip trailing zeros after decimal
    s = repr(value)
    if "e" in s or "E" in s:
        # Normalize scientific notation to lowercase e, strip leading zeros in exponent
        return _normalize_exp(s)
    if "." in s:
        # Strip trailing zeros but keep at least one decimal digit
        s = s.rstrip("0").rstrip(".")
        if "." not in s:
            return s
        return s
    return s


def _normalize_exp(s: str) -> str:
    """Normalize Python float scientific notation to JCS format."""
    # Python uses e+06 style; JCS uses e+6
    import re
    s = s.lower()
    s = re.sub(r"e([+-])0*(\d+)", lambda m: f"e{m.group(1)}{m.group(2)}", s)
    return s

File: sdap-typescript/src/crypto/canonicalize.ts
/**
 * JSON Canonical Serialization (JCS) — RFC 8785.
 */

/**
 * Return RFC 8785 canonical JSON bytes for obj.
 *
 * Rules:
 * - Keys sorted lexicographically (Unicode code-point order), recursively
 * - No insignificant whitespace
 * - Strings encoded as UTF-8 with standard JSON escaping
 * - Numbers: no trailing zeros in fractions; integers have no decimal point
 * - null → null, booleans → true / false
 */
export function canonicalize(obj: Record<string, unknown>): Uint8Array {
  return new TextEncoder().encode(serialize(obj));
}

function serialize(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!isFinite(value)) {
      throw new Error("JCS does not support NaN or Infinity");
    }
    return serializeNumber(value);
  }
  if (typeof value === "string") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    const items = value.map((v) => serialize(v)).join(",");
    return `[${items}]`;
  }
  if (typeof value === "object" && value !== null) {
    const obj = value as Record<string, unknown>;
    const pairs = Object.keys(obj)
      .sort()
      .map((k) => `${JSON.stringify(k)}:${serialize(obj[k])}`)
      .join(",");
    return `{${pairs}}`;
  }
  throw new Error(`Unsupported type for JCS: ${typeof value}`);
}

function serializeNumber(value: number): string {
  // Integer check
  if (Number.isInteger(value)) {
    return value.toString();
  }
  // Use toPrecision for float representation following ES2019 / RFC 8785
  // JavaScript's default number serialization is consistent with ES2019
  const s = String(value);
  if (s.includes("e") || s.includes("E")) {
    return normalizeExp(s);
  }
  return s;
}

function normalizeExp(s: string): string {
  // Normalize scientific notation: e+06 -> e+6
  return s.toLowerCase().replace(/e([+-])0*(\d+)/, (_, sign: string, digits: string) => `e${sign}${digits}`);
}
